src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/LocalFolderUtility.cs

```diff
@@ -0,0 +1,753 @@
+            }
+
+            if (file == null)
+            {
+                throw new ArgumentNullException(nameof(file));
+            }
+
+            return GetIdentityFromNupkgPath(file, id) != null;
+        }
+
+        /// <summary>
+        /// An imperfect attempt at finding the identity of a package from the file name.
+        /// This can fail if the package name ends with something such as .1
+        /// </summary>
+        public static PackageIdentity GetIdentityFromNupkgPath(FileInfo file, string id)
+        {
+            if (id == null)
+            {
+                throw new ArgumentNullException(nameof(id));
+            }
+
+            if (file == null)
+            {
+                throw new ArgumentNullException(nameof(file));
+            }
+
+            var version = GetVersionFromFileName(file.Name, id, PackagingCoreConstants.NupkgExtension);
+
+            if (version != null)
+            {
+                return new PackageIdentity(id, version);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// An imperfect attempt at finding the version of a package from the file name.
+        /// This can fail if the package name ends with something such as .1
+        /// </summary>
+        public static NuGetVersion GetVersionFromFileName(string fileName, string id, string extension)
+        {
+            if (id == null)
+            {
+                throw new ArgumentNullException(nameof(id));
+            }
+
+            if (fileName == null)
+            {
+                throw new ArgumentNullException(nameof(fileName));
+            }
+
+            if (extension == null)
+            {
+                throw new ArgumentNullException(nameof(extension));
+            }
+
+            NuGetVersion result = null;
+            var prefix = $"{id}.";
+
+            if (fileName.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)
+                && fileName.EndsWith(extension, StringComparison.OrdinalIgnoreCase))
+            {
+                fileName = Path.GetFileNameWithoutExtension(fileName);
+
+                // Skip symbol packages
+                if (fileName.Length > prefix.Length && !fileName.EndsWith(".symbols", StringComparison.OrdinalIgnoreCase))
+                {
+                    var versionString = fileName.Substring(prefix.Length, fileName.Length - prefix.Length);
+
+                    NuGetVersion version;
+                    if (NuGetVersion.TryParse(versionString, out version))
+                    {
+                        result = version;
+                    }
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Retrieve a single package from a v3 version folder.
+        /// </summary>
+        public static LocalPackageInfo GetPackageV3(string root, string id, NuGetVersion version, ILogger log)
+        {
+            if (root == null)
```

Comment:
root and log are validated in the other overload. I understand putting them here for completeness. Just thought I'd flag it. Personally, I'd be inclined to put a comment indicating the validation for these is elsewhere, especially because the overload is right below. Your call--just a perf thought.